documentation/man/man1/setprem.1
.TH SETPREM 1 "March 1, 2021" "University of Utrecht"
.UC 4
.SH NAME
setprem \- uses element center coordinates to assign radial elasticity variations to elements following the classical Preliminary Reference Earth Model (PREM) of Dziewonsky and Anderson (1981).
.SH SYNOPSIS
\fBsetprem\fP
.SH DESCRIPTION
\fBsetprem\fP is a support program for finite element code \fIGTECTON\fP, and specifically \fBf3d\fP,
to assist in defining elastic properties in elements. Node and element file 
names are assumed to be tecin.dat.nps and tecin.dat.elm. The third node 
coordinate is assumed to represent the vertical distance with respect to the 
free surface. \fBsetprep\fP writes file tecin.dat.elasticity, with the first 
record refering to the first element, the second to the second element and so 
on. The first column is the element material number, second column Young's modulus (Pa) and third column is Poisson's ratio. 
.SH SEE ALSO
\fIf3d(1)\fP, \fIf3d(7)\fP
.SH AUTHORS
Rob Govers


Changelog
Version     Date                 FEM/PLT   Change

1.4         1989                   ALL     Original version
1.51        September 1991         FEM     Updated plastic code, residual loads.
2.00        October/November 1991  FEM     Added thermal code.

2.01        March 1992             FEM     Debugged thermo-mechanical version
                                           Diffusion creep code.
2.02        August 1992            FEM     Linked nodes/no-tilt boundaries
2.03        October 7 1992         FEM     Consistent plane strain thermal expansion loads.
2.04        November 2 1992        FEM     Solution to diffusion eqn for surface uplifts
                                           and sedimentary transport.
2.05        November 6 1992        FEM     ITMODE=2 (initial temperatures)

2.06        June 7 1993            FEM     Array bounds checking
                                   FEM     Gravity using Talwani's method
2.07        July 1 1993            FEM     SURF arrays always read in. NSED=switch.
2.08        July 13 1993           FEM     Cohesion calculation based upon actual surface
                                           height
2.09        August 20 1993         FEM     Cohesion factor definition changed to include
                                           density
2.10        November 1993          FEM     Slippery nodes update (dimensioning)

2.10        March 1994             FEM     ICVIS switch added
2.11        June 1994              FEM     Thermal consequences of faults
2.12        November 1994          FEM     Sparse solvers (Petsc)

2.12        October 1995           FEM     Bug in ADDSN for triangles

2.13        May 1997               FEM     Incompatible modes and fault intersections

98.1        April 28 1998          ALL     Various modifications
                                   FEM     Out-of-plane library

99.1i       May 1999               FEM     Iterative solvers PetSc

2000        August 2000            FEM     Spherical shell elements

2001        July 2001              FEM     Consistent stress invariants calculations
                                   FEM     Restart facility

2003        June 2003              FEM     Tetrahedral elements

2005        November 2005          FEM     Petsc-2.2.1

2006        June 2006              PLT     GMT Support    

2008        January 2008           FEM     Incompressible plastic flow

2009        June 2009              FEM     Multiple output files (one per time step)

2010        April 2010             FEM     Incorporation Petsc-2.3.2 or higher
                                   FEM     Enabling of parallel computing
                                   FEM     Enhanced sparsity pattern usage for parallel computing

2013        May 2013               FEM     Multinode code runs
                                   ALL     Filepaths no longer restricted to 80 chars. In stead to 255

2014        February 2014          ALL     Upgraded to Fortran 2008 standard
                                   FEM     PETSc 3.4.2

2015        March 2015             ALL     Added axisymmetric solver and plotting

2016        September 2016         ALL     Various bugs

2017.3      March 2017             ALL     Slippery nodes working on multiptocessor runs.
                                           Most of the implicit variables removed.
                                           Part of the static memory allocation removed.

2017.9      September 2017         ALL     Polysort has maximum polygon size increased from 10k to 10M
                                           bugfix: plnplt 'slip' now gives good coordinates
                                           bugfix: plnplt 'nbc' command now returns values.
                                           bugfix: FEDSK files now contain correct velocities in ascii,
                                           More implicitness removed, implicitly fixing bugs.
                                           bugfix: Plt3d now correctly computes centers of elements.
                                           bugfix: Plnplt now correctly writes strain rate files (rate command)
                                           bugfix: ...and elemenental boundary condition file (ebc command)
                                           plnplt now accepts blank lines in scripts
                                           plnplt and plt3d help now take '? set' beside '? se'
                                           elmside now gives an error if node or elem file is empty
                                           Now use system soelim. Minimum version 1.21, 
                                                 in stead of own implementation
                                           bugfix: Elmside now scans element array with element index,
                                                 and nodal point arrays with nodal point index, instead
                                                 of both with the element index.
                                           The main storage is now dynamically allocated. This limits
                                                 the data size to the machine memory size, in stead of to
                                                 the 2GB maximum. These allocations can be set in module 
                                                 TempSrc/AAAB.F
                                           bugfix: The content of environment variable MYTMP is now used as 
                                                 temporary directory. Somehow this feature broke down in the
                                                 past.
                                           More implicit variables removed.
                                           Plt3d has the 'edi' command fixed for total, x, y and z 
                                                 displacements. Slippery and split displacements remain

src/01_meshing/3D/gmsh-helpers/setprem.f90
program setmaterials

implicit none
character (len=1) :: a
integer :: NUMNP,i,node,label
double precision, allocatable :: X(:,:),XELM(:,:)
double precision :: E,v
integer :: NUMEL,element
integer, allocatable :: MAT(:),IEN(:,:)

interface 
    subroutine PREM(MAT,XELM,E,v)
        integer, intent(in) :: MAT
        double precision, intent(in) :: XELM(3)
        double precision, intent(out) :: E,v
    end subroutine PREM

    subroutine ELCTR(X,IEN,XELM)
        integer, intent(in) :: IEN(4)
        double precision, intent(in) :: X(3,*)
        double precision, intent(out) :: XELM(3)
    end subroutine
end interface

open (unit=10,file='tecin.dat.nps',status='old')
rewind (10)

i = 0
do
   read(10,'(a1)',end=100,err=1000) a
   i = i + 1
enddo
100 NUMNP = i-1
allocate (X(3,NUMNP))
rewind(10)
do i=1,NUMNP
    read(10,1,err=1000,end=1100) node,label,X(1,i),X(2,i),X(3,i)
1   format(2I12,3E25.0)
enddo
close(10)

open (unit=10,file='tecin.dat.elm',status='old')
rewind (10)
i = 0
do
   read(10,'(a1)',end=200,err=2000) a
   i = i + 1
enddo
200 NUMEL = i-1
allocate (IEN(4,NUMEL))
allocate (MAT(NUMEL))
rewind(10)
do i=1,NUMEL
    read(10,2,err=2000,end=2100) element,MAT(i),IEN(1,i),IEN(2,i),IEN(3,i),IEN(4,i)
2   format(6I12)
enddo
close(10)

allocate (XELM(3,NUMEL))
do i=1,NUMEL
    call ELCTR(X,IEN(1,i),XELM(1,i))
enddo

deallocate (X)
deallocate (IEN)

open (unit=10,file='tecin.dat.elasticity')
rewind(10)

! assign element property in subroutine (potentially based on material number)
do i=1,NUMEL
    call PREM(MAT(i),XELM(1,i),E,v)
    write(10,3) MAT(i),E,v
3   format(I12,2E25.12)
enddo

close(10)
deallocate (MAT)
deallocate (XELM)
stop

1000 write(0,*) 'setmaterials: read error tecin.dat.nps record=',i
stop
1100 write(0,*) 'setmaterials: EOF error tecin.dat.nps'
stop
2000 write(0,*) 'setmaterials: read error tecin.dat.elm record=',i
stop
2100 write(0,*) 'setmaterials: EOF error tecin.dat.elm'
stop
end program
!-------------------------------------------------------------------------------
subroutine PREM(MAT,XELM,E,v)

! Compute elastic Young's modulus and Poisson ratio from PREM. Material number is irrelevant.
! Assumption is that the third argument of XELM represents distance (height or depth)
! in meters from the free surface.

implicit none
integer, intent(in) :: MAT
double precision, intent(in) :: XELM(3)
double precision, intent(out) :: E,v    ! Young's modulus (Pa) and Poisson ratio

integer :: i
double precision :: Depth(13),BulkModulus(13),ShearModulus(13)
double precision :: distance,dKdz,Ki,dGdz,Gi

! initialize PREM depth (km), K (GPa), G (GPa)
Depth = (/ 0d0, 12d0, 40d0, 68d0, 170d0, 200d0, 220d0, 271d0, 371d0, 400d0, &
  471d0, 571d0, 670d0 /) 
BulkModulus = (/ 52d0, 75.3d0, 131.5d0, 130.4d0, 128.1d0, 128.1d0, 152.9d0, &
  158.6d0, 170.1d0, 189.9d0, 209.7d0, 239.7d0, 255.6d0 /)
ShearModulus = (/ 26.6d0, 44.1d0, 68.2d0, 67.6d0, 66.2d0, 66.2d0, 74.1d0, &
  75.9d0, 79.5d0, 90.6d0, 100.7d0, 116.2d0, 123.9d0 /)

distance = ABS(XELM(3)*1D-3)

i = 1
do while (i.lt.13)
    if ( (Depth(i)-distance)*(Depth(i+1)-distance).le.0d0 ) then
        dKdz = (BulkModulus(i+1)-BulkModulus(i))/(Depth(i+1)-Depth(i))
        Ki = BulkModulus(i) + dKdz*(distance-Depth(i))
        dGdz = (ShearModulus(i+1)-ShearModulus(i))/(Depth(i+1)-Depth(i))
        Gi = ShearModulus(i) + dGdz*(distance-Depth(i))
        E = 9d0*Ki*Gi/(3d0*Ki+Gi) * 1d9
        v = (3d0*Ki-2d0*Gi)/(6d0*Ki+2d0*Gi)
        return
    endif
    i = i + 1
enddo

write(0,*) 'PREM: ',distance,' km not in table'
stop

end subroutine
!-------------------------------------------------------------------------------
subroutine ELCTR(X,IEN,XELM)

! compute element center coordinate

implicit none
integer, intent(in) :: IEN(4)
double precision, intent(in) :: X(3,*)
double precision, intent(out) :: XELM(3)
double precision ::XL(3,4)
integer :: k

k = IEN(1)
XL(1,1) = X(1,k)
XL(2,1) = X(2,k)
XL(3,1) = X(3,k)
k = IEN(2)
XL(1,2) = X(1,k)
XL(2,2) = X(2,k)
XL(3,2) = X(3,k)
k = IEN(3)
XL(1,3) = X(1,k)
XL(2,3) = X(2,k)
XL(3,3) = X(3,k)
k = IEN(4)
XL(1,4) = X(1,k)
XL(2,4) = X(2,k)
XL(3,4) = X(3,k)
XELM(1) = (xl(1,1)+xl(1,2)+xl(1,3)+xl(1,4))*0.25d0
XELM(2) = (xl(2,1)+xl(2,2)+xl(2,3)+xl(2,4))*0.25d0
XELM(3) = (xl(3,1)+xl(3,2)+xl(3,3)+xl(3,4))*0.25d0

return
end subroutine
